src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs

```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
         IOptionsMonitor<EventPropagationSubscriptionOptions> optionsMonitor,
         ILogger<EventPullerService> logger)
     {
-        this._serviceScopeFactory = serviceScopeFactory;
-        this._logger = logger;
-        this._eventGridTopicSubscriptions = clientDescriptors.Select(descriptor =>
-            new EventGridTopicSubscription(
-                optionsMonitor.Get(descriptor.Name).TopicName,
-                optionsMonitor.Get(descriptor.Name).SubscriptionName,
-                eventGridClientWrapperFactory.CreateClient(descriptor.Name))).ToArray();
+        this._eventGridSubscriptionPullers = clientDescriptors.Select(descriptor =>
+                new EventGridSubscriptionEventPuller(
+                    new EventGridTopicSubscription(
+                        optionsMonitor.Get(descriptor.Name).TopicName,
+                        optionsMonitor.Get(descriptor.Name).SubscriptionName,
+                        optionsMonitor.Get(descriptor.Name).MaxDegreeOfParallelism,
+                        eventGridClientWrapperFactory.CreateClient(descriptor.Name)),
+                    serviceScopeFactory,
+                    logger))
+            .ToList();
     }
 
     protected override Task ExecuteAsync(CancellationToken stoppingToken)
     {
-        return Task.WhenAll(this._eventGridTopicSubscriptions.Select(sub => Task.Run(() => this.StartReceivingEventsAsync(sub, stoppingToken), stoppingToken)));
+        return Task.WhenAll(this._eventGridSubscriptionPullers.Select(puller => Task.Run(() => puller.StartReceivingEventsAsync(stoppingToken), stoppingToken)));
     }
 
-    private async Task StartReceivingEventsAsync(EventGridTopicSubscription eventGridTopicSubscription, CancellationToken stoppingToken)
+    private class EventGridSubscriptionEventPuller
     {
-        using var scope = this._serviceScopeFactory.CreateScope();
-        var cloudEventHandler = scope.ServiceProvider.GetRequiredService<ICloudEventHandler>();
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+        private readonly ILogger<EventPullerService> _logger;
+        private readonly EventGridTopicSubscription _eventGridTopicSubscription;
 
-        while (!stoppingToken.IsCancellationRequested)
+        private readonly Channel<Task<EventHandlingResult>> _handlerTasksChannel;
+        private readonly Channel<string> _acknowledgeEventChannel = Channel.CreateUnbounded<string>();
```

Comment:
I think these channels should be bounded. Most of the time it should not be a problem, but if you process messages to quickly, or if there is an issue, it can create back-pressure. Also processed messages may timeout and will be reprocessed.
A bounded channel may be a safety-net

Reply:
Switched to bounded channels with a limit of 5000.